libgeopmd/src/DrmGpuTopo.cpp

```diff
@@ -0,0 +1,247 @@
+                                   GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+        int next_segment_cpu_offset = cpu_offset + MAX_CPUS_PER_CPUMASK_SEGMENT;
+        while (cpu_offset < next_segment_cpu_offset) {
+            if (bitmask_segment & 1) {
+                mapped_cpus.insert(cpu_offset);
+            }
+            bitmask_segment >>= 1;
+            cpu_offset += 1;
+        }
+    }
+    return mapped_cpus;
+}
+
+// Return the name of the driver that provides the given /sys/class/drm/card*/ device
+static std::string drm_driver_name_from_card_path(const std::string &card_path)
+{
+    auto driver_path = geopm::read_symlink_target(card_path + "/device/driver");
+    return driver_path.substr(driver_path.find_last_of("/") + 1);
+}
+
+static std::vector<std::string> get_file_paths_with_pattern(
+    const std::string &parent_directory_path,
+    const std::regex &pattern)
+{
+    std::vector<std::string> paths;
+    std::vector<std::string> directory_files;
+    try {
+        directory_files = geopm::list_directory_files(parent_directory_path);
+    }
+    catch (const geopm::Exception &e) {
+    }
+
+    for (const auto &child_file : directory_files) {
+        std::smatch file_match;
+        if (!std::regex_match(child_file, file_match, pattern)) {
+            continue;
+        }
+
+        if (file_match.size() != 2) {
+            // There should be two matches: the whole regex and one capture group
+            throw std::logic_error(std::string(__func__) + ": Expected 2 matching groups, found " +
+                                   std::to_string(file_match.size()));
+        }
+        paths.emplace_back(parent_directory_path + "/" + child_file);
+    }
+    return paths;
+}
+
+// Query which driver provides each card. Return the pair of (driver name) and
+// (vector of card drm paths) of the driver with the most cards present.
+using DriverName = std::string;
+using CardVector = std::vector<std::string>;
+using DriverCards = std::pair<DriverName, CardVector>;
+static DriverCards get_cards_from_most_frequent_driver(const CardVector &all_cards)
+{
+    std::map<DriverName, CardVector> driver_card_paths; // Map of (driver name)->(vector of card paths)
+    for (const auto &card_path : all_cards) {
+        auto driver_name = drm_driver_name_from_card_path(card_path);
+        driver_card_paths[driver_name].push_back(card_path);
+    }
+    auto driver_with_max_cards_it = std::max_element(
+        driver_card_paths.begin(), driver_card_paths.end(),
+        [](const DriverCards &lhs, const DriverCards &rhs) {
+            return lhs.second.size() < rhs.second.size();
+        });
+    if (driver_with_max_cards_it == driver_card_paths.end()) {
+        // This should only happen if driver_card_paths (and all_cards) are empty.
+        throw std::logic_error(std::string(__func__) + ": No max-count is defined for driver card paths");
+    }
+    return *driver_with_max_cards_it;
```

Comment:
I think this is returns the driver that has the largest number of cards associated with it.  Maybe a comment about the return value could clarify this.